{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import scanpy as sc\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import argparse\n",
        "import random\n",
        "import numpy as np\n",
        "import mygene\n",
        "\n",
        "\n",
        "def generate_cnv_region(chromosomes, min_size, max_size):\n",
        "\n",
        "    chrom = random.choice(chromosomes)\n",
        "    start = random.randint(1, 100000000) \n",
        "    end = start + random.randint(min_size, max_size)\n",
        "\n",
        "    return chrom, start, end\n",
        "\n",
        "\n",
        "def assign_cnvs_unique(cell_types, num_cnvs, chromosomes, min_size, max_size, cnv_prob=0.1):\n",
        "\n",
        "    cnvs = []\n",
        "    for _ in range(num_cnvs):\n",
        "        chrom, start, end = generate_cnv_region(chromosomes, min_size, max_size)\n",
        "        cnv_type = random.choice(['gain', 'loss'])\n",
        "        cnvs.append({'chromosome': chrom, 'start': start, 'end': end, 'type': cnv_type})\n",
        "\n",
        "    cnv_assignments = {}\n",
        "    used_cnvs = []  # track CNVs that have been assigned to any cell type\n",
        "\n",
        "    for cell_type in cell_types:\n",
        "        cell_indices = adata2.obs[adata2.obs['cell_type'] == cell_type].index\n",
        "        num_cells_in_type = len(cell_indices)\n",
        "        num_cnv_cells = int(np.round(num_cells_in_type * cnv_prob)) \n",
        "\n",
        "        cnv_cells = random.sample(cell_indices.tolist(), num_cnv_cells)\n",
        "\n",
        "        #choose a CNV that hasn't been assigned yet\n",
        "        available_cnvs = [cnv for cnv in cnvs if cnv not in used_cnvs]\n",
        "        if not available_cnvs:\n",
        "            break  \n",
        "        chosen_cnv = random.choice(available_cnvs) \n",
        "        cnv_type_for_cell_type = chosen_cnv['type']\n",
        "\n",
        "        used_cnvs.append(chosen_cnv)\n",
        "        cnv_assignments[cell_type] = {\n",
        "            cell: {'cnv': chosen_cnv, 'type': cnv_type_for_cell_type} for cell in cnv_cells\n",
        "        }\n",
        "    return cnv_assignments\n",
        "\n",
        "\n",
        "def simulate_cnv_impact(adata, cnv_assignments):\n",
        "    for cell_type, cnvs in cnv_assignments.items():\n",
        "        for cell, cnv_info in cnvs.items():\n",
        "            cnv = cnv_info['cnv']  \n",
        "            affected_genes = adata.var[(adata.var['chromosome'] == cnv['chromosome']) &\n",
        "                                       (adata.var['start'] >= cnv['start']) &\n",
        "                                       (adata.var['end'] <= cnv['end'])]\n",
        "            affected_genes_positions = [adata.var_names.get_loc(gene) for gene in affected_genes.index]\n",
        "            if cnv_info['type'] == 'gain':\n",
        "                for gene_idx in affected_genes_positions:\n",
        "                    adata.layers['counts'][:, gene_idx] *= 3  # Scaling up expression for 'gain'\n",
        "            elif cnv_info['type'] == 'loss':\n",
        "                for gene_idx in affected_genes_positions:\n",
        "                    adata.layers['counts'][:, gene_idx] *= 2  # Scaling down expression for 'loss'\n",
        "\n",
        "    return adata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Input sequence provided is already in string format. No operation performed\n",
            "Input sequence provided is already in string format. No operation performed\n",
            "5 input query terms found dup hits:\t[('ENSG00000234162', 2), ('ENSG00000227110', 2), ('ENSG00000249738', 2), ('ENSG00000280018', 2), ('E\n",
            "411 input query terms found no hit:\t['ENSG00000238009', 'ENSG00000230699', 'ENSG00000236948', 'ENSG00000277726', 'ENSG00000271895', 'ENS\n",
            "/Users/vidyaajay/Desktop/cscb/final_project/.venv/lib/python3.10/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
            "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
            "/Users/vidyaajay/Desktop/cscb/final_project/.venv/lib/python3.10/site-packages/scipy/sparse/_data.py:57: RuntimeWarning: overflow encountered in multiply\n",
            "  self.data *= other\n",
            "/var/folders/fq/5tx3gdns3wb_2c22jz7wrp4r0000gn/T/ipykernel_1177/3082197321.py:59: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
            "  cnv_table = adata2_CNV.obs.groupby(['simulated_cnvs', 'cell_type']).size().unstack(fill_value=0)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "cell_type                    B cell  CD14 monocyte  CD4 T cell  CD8 T cell  \\\n",
            "simulated_cnvs                                                               \n",
            "                               1450           3128        3199        1029   \n",
            "20:34285250-34774160 (loss)       0              0         355           0   \n",
            "\n",
            "cell_type                    Dendritic  FCGR3A monocyte  Megakaryocyte  \\\n",
            "simulated_cnvs                                                           \n",
            "                                   154              327             59   \n",
            "20:34285250-34774160 (loss)          0                0              0   \n",
            "\n",
            "cell_type                    NK cell  \n",
            "simulated_cnvs                        \n",
            "                                 608  \n",
            "20:34285250-34774160 (loss)        0  \n"
          ]
        }
      ],
      "source": [
        "\n",
        "SEED = 5  \n",
        "random.seed(SEED)\n",
        "np.random.seed(SEED)\n",
        "\n",
        "adata2 = sc.read_h5ad(\"files/adPBMC_ref_040623.h5ad\")\n",
        "adata2.layers['counts'] = adata2.X.copy()\n",
        "\n",
        "mg = mygene.MyGeneInfo()\n",
        "\n",
        "\n",
        "gene_info = mg.querymany(\n",
        "    adata2.var['gene_ids'].tolist(),\n",
        "    scopes='ensembl.gene',\n",
        "    fields='genomic_pos',\n",
        "    species='human'\n",
        ")\n",
        "\n",
        "\n",
        "query_df = pd.DataFrame(gene_info)\n",
        "query_df['genomic_pos'] = query_df['genomic_pos'].apply(lambda x: eval(x) if isinstance(x, str) else x)\n",
        "\n",
        "#extract chromosome, start, and end info from the genomic_pos column\n",
        "query_df['chromosome'] = query_df['genomic_pos'].apply(lambda x: x.get('chr') if isinstance(x, dict) else None)\n",
        "query_df['start'] = query_df['genomic_pos'].apply(lambda x: x.get('start') if isinstance(x, dict) else None)\n",
        "query_df['end'] = query_df['genomic_pos'].apply(lambda x: x.get('end') if isinstance(x, dict) else None)\n",
        "query_df['strand'] = query_df['genomic_pos'].apply(lambda x: x.get('strand') if isinstance(x, dict) else None)\n",
        "\n",
        "# Remove rows where 'chromosome' is None or 'start' is NaN\n",
        "clean_query_df = query_df.dropna(subset=['chromosome', 'start', 'end'])\n",
        "clean_query_df = clean_query_df.drop_duplicates(subset=['query'])\n",
        "\n",
        "adata2.var = adata2.var.merge(clean_query_df[['query', 'chromosome', 'start', 'end', 'strand']],\n",
        "                              left_on='gene_ids', right_on='query',\n",
        "                              how='left')\n",
        "\n",
        "adata2_CNV = adata2.copy()\n",
        "\n",
        "# Define parameters for CNV simulation\n",
        "chromosomes = [str(i) for i in range(1, 23)] + ['X', 'Y']  \n",
        "min_size = 100000 \n",
        "max_size = 500000  \n",
        "num_cnvs = 1  # Number of CNVs to simulate\n",
        "\n",
        "\n",
        "cell_types = adata2_CNV.obs['cell_type'].unique()\n",
        "\n",
        "\n",
        "cnv_prob = 0.1  # 10% of cells per cell type will receive CNVs\n",
        "cnv_assignments = assign_cnvs_unique(cell_types, num_cnvs, chromosomes, min_size, max_size, cnv_prob)\n",
        "\n",
        "# Simulate CNV impact on expression\n",
        "adata2_CNV = simulate_cnv_impact(adata2_CNV, cnv_assignments)\n",
        "adata2_CNV.obs['simulated_cnvs'] = ''\n",
        "for cell_type, cell_cnv_mapping in cnv_assignments.items():\n",
        "    for cell, cnv_info in cell_cnv_mapping.items():\n",
        "        cnv_desc = f\"{cnv_info['cnv']['chromosome']}:{cnv_info['cnv']['start']}-{cnv_info['cnv']['end']} ({cnv_info['cnv']['type']})\"\n",
        "        adata2_CNV.obs.loc[cell, 'simulated_cnvs'] = cnv_desc\n",
        "\n",
        "cnv_table = adata2_CNV.obs.groupby(['simulated_cnvs', 'cell_type']).size().unstack(fill_value=0)\n",
        "\n",
        "print(cnv_table)\n",
        "adata2_CNV.write('PBMC_simCNV_3.h5ad')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>cell_type</th>\n",
              "      <th>B cell</th>\n",
              "      <th>CD14 monocyte</th>\n",
              "      <th>CD4 T cell</th>\n",
              "      <th>CD8 T cell</th>\n",
              "      <th>Dendritic</th>\n",
              "      <th>FCGR3A monocyte</th>\n",
              "      <th>Megakaryocyte</th>\n",
              "      <th>NK cell</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>simulated_cnvs</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th></th>\n",
              "      <td>1450</td>\n",
              "      <td>3128</td>\n",
              "      <td>3199</td>\n",
              "      <td>1029</td>\n",
              "      <td>154</td>\n",
              "      <td>327</td>\n",
              "      <td>59</td>\n",
              "      <td>608</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20:34285250-34774160 (loss)</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>355</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "cell_type                    B cell  CD14 monocyte  CD4 T cell  CD8 T cell  \\\n",
              "simulated_cnvs                                                               \n",
              "                               1450           3128        3199        1029   \n",
              "20:34285250-34774160 (loss)       0              0         355           0   \n",
              "\n",
              "cell_type                    Dendritic  FCGR3A monocyte  Megakaryocyte  \\\n",
              "simulated_cnvs                                                           \n",
              "                                   154              327             59   \n",
              "20:34285250-34774160 (loss)          0                0              0   \n",
              "\n",
              "cell_type                    NK cell  \n",
              "simulated_cnvs                        \n",
              "                                 608  \n",
              "20:34285250-34774160 (loss)        0  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "cnv_table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": ".venv",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.16"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
